### PR TITLE
11.0 release

### PR DESCRIPTION
**Pull request by kirkbushell**
Currently work in progress, but this PR encapsulates the complete rework of the Eloquence solution, using a much nicer and workable API with modern approaches.

The main objective of version 11 is to prepare for Laravel 11 (in general), whilst introducing a nicer API and easier way of working with the main camelcasing trait, and providing better support for its various data aggregation features. 

Due to the amount of changes, backwards compatibility is not possible.

Upgrade guide to come.

### CODE REVIEW

File: src/Behaviours/CountCache/Observer.php [hunk 1]
```diff
@@ -7,58 +7,55 @@
...
     /**
      * When the model is restored, again increment the count cache by 1.
      *
      * @param $model
      */
-    public function restored($model)
+    public function restored($model): void
     {
-        $this->update($model, '+');
+        CountCache::for($model)->increment($model);
```
Comment (kontoulis) [inline src/Behaviours/CountCache/Observer.php:47]: `increment()` takes no argument

File: src/Behaviours/CountCache/Observer.php [hunk 2]
```diff
@@ -7,58 +7,55 @@
...
-    private function update($model, $operation)
+    private function update($model, string $operation): void
     {
-        $countCache = new CountCache($model);
-        $countCache->apply(function ($config) use ($countCache, $model, $operation) {
-            $countCache->updateCacheRecord($config, $operation, 1, $model->{$config['foreignKey']});
-        });
+        $countCache = CountCache::for($model);
+
+        $countCache->apply((fn($config) => $this->updateCacheRecord($config->relatedModel($model), $config, $operation))->bindTo($countCache));
```
Comment (kontoulis) [inline src/Behaviours/CountCache/Observer.php:59]: This method is not used :thinking:

File: src/Behaviours/Sluggable.php [hunk 3]
```diff
@@ -63,27 +63,22 @@ public function generateTitleSlug(array $fields)
...
-        return $fields;
     }
 
     /**
      * Generate the slug for the model based on the model's slug strategy.
      */
-    public function generateSlug()
+    public function generateSlug(): void
     {
         $strategy = $this->slugStrategy();
```
Comment (kontoulis) [inline src/Behaviours/Sluggable.php:84]: Is there a pending implementation for `Uuid` ?
I see Ramsey's package being added in the `composer.json`

File: src/Behaviours/SumCache/SumCache.php [hunk 4]
```diff
@@ -1,127 +1,79 @@
...
-     *
-     * @param \Closure $function
-     */
-    public function apply(\Closure $function)
-    {
-        foreach ($this->model->sumCaches() as $key => $cache) {
-            $function($this->config($key, $cache));
-        }
+        return $this->reflect(SummedBy::class, function(array $config) {
+            return [$config['name'] => [$config['attribute']->as => $config['attribute']->from]];
```
[the remark this answers is not in the document]
Comment (kirkbushell) [reply src/Behaviours/SumCache/SumCache.php:18]: @kontoulis On your first point - yes, but that also makes the structure harder to work with when it's consumed, thereby making other code harder to read. This is the output:

array:1 [
  "order" => array:1 [
    "total_amount" => "amount"
  ]
]

So it's easy to go - order relationship, the field, and the other field.

File: src/Behaviours/Cacheable.php [hunk 5]
```diff
@@ -1,144 +1,110 @@
...
-        if ($operation == '+') {
-            return $sql->increment($config['field'], $amount);
-        }
-
-        /*
-         * Decrement for - operator
-         */
-        return $sql->decrement($config['field'], $amount);
-    }
+    abstract private function configuration(): array;
```
[the remark this answers is not in the document]
Comment (kirkbushell) [reply src/Behaviours/Cacheable.php:21]: It's not used outside of the classes that use the Cacheable trait, though.

File: src/Behaviours/Sluggable.php (hunk 3 above)
Comment (kirkbushell) [reply src/Behaviours/Sluggable.php:84]: There is not, now been removed.

**Comment by kirkbushell** [comment]
Moving to release branch.

**Pull request closed without merging**